frontend: convert chart.tsx component

NicolaLS opened the pull request:
Convert the Chart class component in chart.tsx to a functional component.


There some general improvements that could be made, but this PR aims to perform minimal changes to only transform the component from a class component to a functional one. I will open another PR with a general refactor when this one was merged.

note: I moved the `renderDate` function out of the component, but this improvement should've been in the next PR not this one, lmk if I should put it back.

thisconnect commented on frontends/web/src/routes/account/summary/chart.tsx:
```diff
@@ -49,120 +43,176 @@ type State = {
-
-  public componentDidMount() {
-    this.createChart();
+  });
```
split into source/setSource, difference/setDifference, diffSince/setDiffSince and the rest tooltip/setTooltip ?

thisconnect commented on frontends/web/src/routes/account/summary/chart.tsx:
```diff
@@ -49,120 +43,176 @@ type State = {
+  const [prevChartFiat, setPrevChartFiat] = useState(data.chartFiat);
+  const [prevHideAmounts, setPrevHideAmounts] = useState(hideAmounts);
+
+  if (prevChartDataDaily !== data.chartDataDaily) {
```
could hooks/usePrevious be helpful here?

thisconnect commented on frontends/web/src/routes/account/summary/chart.tsx:
```diff
@@ -314,292 +362,237 @@ class Chart extends Component<Props, State> {
       from,
     };
-  };
+  }
```
looks like this function could be moved out of the component next to renderDate

NicolaLS commented:
addressed everything, PTAL @thisconnect :)

thisconnect commented on frontends/web/src/routes/account/summary/chart.tsx:
```diff
@@ -28,141 +29,195 @@ import { DefaultCurrencyRotator } from '../../../components/rates/rates';
-      this.lineSeries.setData(data);
-      this.chart?.timeScale().fitContent();
-      this.setFormattedData(data);
+  });
```
I believe this effect is not needed, as the previous effect already and unconditionally calls initChart().

NicolaLS replied:
yes, I also thought that, but I thought I should put it in there since there was [this](https://github.com/BitBoxSwiss/bitbox-wallet-app/pull/2728/files#diff-c761d5505dba4edd4496dfe892f927231782977bae81a155d645a6a04dd7ab59L109).

So yes logically it does not need to be there but if we want to strictly say we only want to convert the component it should be there? But I'll remove it and test if everything still works.

thisconnect commented on frontends/web/src/routes/account/summary/chart.tsx:
```diff
@@ -28,141 +29,195 @@ import { DefaultCurrencyRotator } from '../../../components/rates/rates';
+        chart.current.unsubscribeCrosshairMove(handleCrosshair);
+      }
+    };
+  // eslint-disable-next-line react-hooks/exhaustive-deps
```
is this needed?

NicolaLS replied:
we explicitly want to run the effect **once** which is possible by passing an empty dependency array according to the react docs. however eslint wants us to either add the used dependencies, or remove the dependency array (which would cause it to run every render). 

so we either ignore the warning with the comment or remove the dependency array but check if it already ran once e.g
```
const chartInitialized = useRef(false);

useEffect..
  if(!chartInitialized) ...chartInitialized = true, initChart() return cleanup
  else do nothing
 ```
 
 so the function in use effect would return `() => void | undefined` not sure if thats even possible ? can it have _no cleanup function_ when some condition is not met?

NicolaLS replied:
tldr; yes otherwise we get eslint errors

NicolaLS commented on frontends/web/src/routes/account/summary/chart.tsx:
```diff
@@ -231,59 +286,57 @@ class Chart extends Component<Props, State> {
 
-  private onResize = () => {
-    const isMobile = window.innerWidth <= 640;
-    this.setState({ isMobile });
```
because it is calculated every time the component renders now. see [here](https://github.com/BitBoxSwiss/bitbox-wallet-app/pull/2728/files#diff-c761d5505dba4edd4496dfe892f927231782977bae81a155d645a6a04dd7ab59R88).

edit: ah, the component does not re-render automatically when the window re-sizes..good catch. I will convert `isMobile` to state again, and initialize it just like right now, but also `setIsMobile` in `onResize` then.

thisconnect replied:
can you past full links instead of "here", in some other comment you have like 3 links "here" "here" "this" this is hard for me to read.

also this should not update on ever render but on resize.

thisconnect commented:
> componentDidMount and componentWillUnmount are replaced by [this](https://github.com/BitBoxSwiss/bitbox-wallet-app/pull/2728/files#diff-c761d5505dba4edd4496dfe892f927231782977bae81a155d645a6a04dd7ab59R140).

I believe "this" and a few other links broke. also it's hard to follow please consider using less links or paste full links. but if you rebase / squash etc. such links might break.

NicolaLS commented on frontends/web/src/routes/account/summary/chart.tsx:
```diff
@@ -28,141 +29,195 @@ import { DefaultCurrencyRotator } from '../../../components/rates/rates';
-  private setFormattedData(data: ChartData) {
-    this.formattedData = {} as FormattedData;
+  function setFormattedData(chartData: ChartData) {
+    formattedData.current = {} as FormattedData;
```
removed `as FormattedData`, `as ChartData`, and ` as TChartFiltersProps`

NicolaLS commented:
> > componentDidMount and componentWillUnmount are replaced by [this](https://github.com/BitBoxSwiss/bitbox-wallet-app/pull/2728/files#diff-c761d5505dba4edd4496dfe892f927231782977bae81a155d645a6a04dd7ab59R140).
> 
> I believe "this" and a few other links broke. also it's hard to follow please consider using less links or paste full links. but if you rebase / squash etc. such links might break.

ah I see..I think I will just remove the note now, thanks for letting me know. it was not super important just meant as a quick overview before starting the review but you already did that. (the links just pointed to the respective `useEffect`s..)

NicolaLS commented:
> tested [b95ae02](https://github.com/BitBoxSwiss/bitbox-wallet-app/commit/b95ae02c0b9347e753ed643957dfb8fbd06e39a7) webdev/servewallet and initChart is called, but somehow the chart isn't loading.
> <img alt="Screenshot 2024-05-30 at 11 13 54" width="1221" src="https://private-user-images.githubusercontent.com/546900/335120123-0bc6d471-da29-4163-85a0-df34da5a926e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTcwNzM3ODYsIm5iZiI6MTcxNzA3MzQ4NiwicGF0aCI6Ii81NDY5MDAvMzM1MTIwMTIzLTBiYzZkNDcxLWRhMjktNDE2My04NWEwLWRmMzRkYTVhOTI2ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUzMFQxMjUxMjZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yNWFiN2JmN2NiZmVkYjA3NjU4OWZlY2NlNTZmZTY4ZmEzZjNiYmI5Zjc1NjhkMTdmZjUyMGI4NTg3NzBkMzE2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.6aLDHWkI9ZN4sAeFbpref4x1GJtKHC65rkmcYP0rCus">

actually, I forgot testing after removing the `useEffect` that calls `initChart` if `!chart.current`..but I tested right now without changing anything and `webdev/servewallet` seem to work fine for me on firefox and chrome..I'll look into this though.

NicolaLS commented:
(rebase master force push )

NicolaLS commented:
> tested [b95ae02](https://github.com/BitBoxSwiss/bitbox-wallet-app/commit/b95ae02c0b9347e753ed643957dfb8fbd06e39a7) webdev/servewallet and initChart is called, but somehow the chart isn't loading.
> <img alt="Screenshot 2024-05-30 at 11 13 54" width="1221" src="https://private-user-images.githubusercontent.com/546900/335120123-0bc6d471-da29-4163-85a0-df34da5a926e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTcwNzM3ODYsIm5iZiI6MTcxNzA3MzQ4NiwicGF0aCI6Ii81NDY5MDAvMzM1MTIwMTIzLTBiYzZkNDcxLWRhMjktNDE2My04NWEwLWRmMzRkYTVhOTI2ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUzMFQxMjUxMjZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yNWFiN2JmN2NiZmVkYjA3NjU4OWZlY2NlNTZmZTY4ZmEzZjNiYmI5Zjc1NjhkMTdmZjUyMGI4NTg3NzBkMzE2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.6aLDHWkI9ZN4sAeFbpref4x1GJtKHC65rkmcYP0rCus">

@thisconnect so it is definitely the `useEffect` I removed, I added it back in and the chart shows again. I would argue that for a transformation of the code it is also correct that it is in there, since the original code also contains:
```ts
public componentDidUpdate(prev: Props) {
    const { chartDataDaily, chartDataHourly } = this.props.data;
    if (!this.chart) {
      this.createChart();
    }
```

which translates to the `useEffect` that runs every render. I agree that it should not be like this, but would it make sense to refactor this in another PR? Of course I can do it in this PR too but I think it would be better in another PR.

Do you have any env variables for react? Maybe the chart shows for me because react runs `useEffect` twice in dev mode (so you're not in dev mode?)

thisconnect commented on frontends/web/src/api/account.ts:
```diff
@@ -14,8 +14,8 @@
  * limitations under the License.
  */
 
+import { LineData } from 'lightweight-charts';
```
```suggestion
import type { LineData } from 'lightweight-charts';
```

thisconnect commented on frontends/web/src/routes/account/summary/chart.tsx:
```diff
@@ -15,157 +15,215 @@
-    },
-    hideAmounts: false,
+
+function getUTCRange() {
```
please use arrow functions if possible, I recently changed other functions see https://github.com/BitBoxSwiss/bitbox-wallet-app/pull/2731

NicolaLS replied:
Good to know, so I'll always use them now. I was really confused on how to decide which ones to use..I kept the normal functions so the `git diff` is better because they don't need to be moved around (hoisting)

NicolaLS replied:
actually typescript did not warn me when I changed this, had to move the functions around quite a bit so the order is correct and I did not get any runtime errors...

thisconnect commented on frontends/web/src/routes/account/summary/chart.tsx:
```diff
@@ -15,157 +15,215 @@
   };
+}
+
+function renderDate(date: number, lang: string, src: string) {
```
please use newlines if there is more than 1-2 simple props.

thisconnect commented on frontends/web/src/routes/account/summary/chart.tsx:
```diff
@@ -475,134 +478,109 @@ class Chart extends Component<Props, State> {
+    toolTipTime,
+  } = tooltipData;
+
+
```
nit: for readability and consistency there is no need to have multiple newlines, here and before

thisconnect commented on frontends/web/src/routes/account/summary/chart.tsx:
```diff
@@ -15,157 +15,215 @@
+import { useContext, useEffect, useRef, useState } from 'react';
+import { useTranslation } from 'react-i18next';
+import { usePrevious } from '../../../hooks/previous';
+import { ISummary, ChartData } from '../../../api/account';
```
```suggestion
import type { ISummary, ChartData } from '../../../api/account';
```

thisconnect commented on frontends/web/src/routes/account/summary/chart.tsx:
```diff
@@ -15,157 +15,215 @@
-import { Component, createRef, ReactChild } from 'react';
-import { ISummary } from '../../../api/account';
-import { translate, TranslateProps } from '../../../decorators/translate';
+import { useContext, useEffect, useRef, useState } from 'react';
```
nit: would be nice if react import is always first.

thisconnect commented on frontends/web/src/routes/account/summary/chart.tsx:
```diff
@@ -15,157 +15,215 @@
     }
+    initChart();
+    return () => {
+      window.removeEventListener('resize', onResize);
```
initChart conditionally adds the event listener, but here it always removes the event.

I think this should be changed somehow. I guess the event can always be added?

thisconnect reviewed (changes requested):
found a weird issue, something seems wrong:

- click week 
- rotate to BTC
- white screen & js error

<img width="780" alt="Screenshot 2024-06-07 at 09 19 21" src="https://github.com/BitBoxSwiss/bitbox-wallet-app/assets/546900/923eb93f-7c60-456b-9485-4a8cf4c98a4c">

NicolaLS commented:
> found a weird issue, something seems wrong:
> 
>     * click week
> 
>     * rotate to BTC
> 
>     * white screen & js error
> 
> 
> <img alt="Screenshot 2024-06-07 at 09 19 21" width="780" src="https://private-user-images.githubusercontent.com/546900/337565973-923eb93f-7c60-456b-9485-4a8cf4c98a4c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc3NzE4NjUsIm5iZiI6MTcxNzc3MTU2NSwicGF0aCI6Ii81NDY5MDAvMzM3NTY1OTczLTkyM2ViOTNmLTdjNjAtNDU2Yi05NDg1LTRhOGNmNGM5OGE0Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjA3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYwN1QxNDQ2MDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01Y2Y1OGM1MjcxNmI4ZWMwMmQxNTJjMWQ2Y2U3NjE0OWM0OTQ2ODExYTllNjI2NWQxMzRlYTgwMjhiYzA5OGFlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.yGYJC6nto1k0Um5A_JmsrOdih2F2pG1V4EKvvIEurcE">

ah yes, this also happens when clicking on some account now and then going back to the account overview..its weird because I'm pretty sure that this was not an issue and some point, I don't know what change caused this (maybe it was an issue, I'm not hundred percent sure)

- the problem is we set the state of `chartDisplay` and `source` when the switch in `initChart` matches anything (this is why there is no issue if 'all' is selected. this causes the `chart.current` to stay uninitialized causing the infinite loop.
I will just fix the chart initialization now, thought it would be better in another PR but it seems like I can only get this one to work with a proper chart initialization logic.

- I changed everything in the first commit but it stays broken with the issue you described here
- I will add another commit fixing the initialization, but only for a easy review, the commits need to be squashed since the first one is broken

NicolaLS commented on frontends/web/src/routes/account/summary/chart.tsx:
```diff
@@ -209,10 +210,13 @@ export const Chart = ({
-      chart.current = undefined;
+    if (chartInitialized.current) {
       window.removeEventListener('resize', onResize);
+      chart.current?.timeScale().unsubscribeVisibleLogicalRangeChange(calculateChange);
+      chart.current?.unsubscribeCrosshairMove(handleCrosshair);
```
imo. `chart.remove()` should clean up everything but from a quick look at the function documentation I don't know if it does so I left the unsubscribing in here just for safety..wdyt @thisconnect should I remove these?

NicolaLS commented on frontends/web/src/routes/account/summary/chart.tsx:
```diff
@@ -101,10 +101,11 @@ export const Chart = ({
   const ref = useRef<HTMLDivElement>(null);
   const refToolTip = useRef<HTMLSpanElement>(null);
   const chart = useRef<IChartApi>();
+  const chartInitialized = useRef(false);
```
I guess we could also use `chart.current` but I like it like that, it is more clear even if its a bit redundant. but no strong opinion, let me know if I should remove this and use `chart.current` (!== undefined) to check if the chart was initialized.

NicolaLS reviewed (commented):
@thisconnect I addressed all changes in the first commit and added a second one that refactors the chart initialization. When everything is fine I can squash them before merge. 
PTAL :)

NicolaLS commented on frontends/web/src/routes/account/summary/chart.tsx:
```diff
@@ -475,134 +293,280 @@ class Chart extends Component<Props, State> {
+      <header>
+        <div className={styles.summary}>
+          <div className={styles.totalValue}>
+            {formattedChartTotal !== null ? (
```
note: can be `undefined` needs to be fixed but wait for master and rebase

thisconnect replied:
did you rebase?

NicolaLS replied:
Yes I did rebase, `formattedChartTotal` should now be either `null` or some value because #2850 now uses `success` and `error` response.

unrelated to this: I saw some other comments above, I thought I did address all change requests/review but will take another look and ping you if I think it is ready for review again, thanks :)

NicolaLS commented:
@thisconnect all change requests should be fixed/outdated now, I did not respond to all the comments individually to create less noise. PTAL or resolve the threads that are satisfied in case there is more to do (thanks :))

NicolaLS commented:
(rebased without changes)

NicolaLS commented on frontends/web/src/routes/account/summary/chart.tsx:
```diff
@@ -433,45 +425,36 @@ export const Chart = ({
+      removeChart();
     };
     // eslint-disable-next-line react-hooks/exhaustive-deps
   }, []);
```
I can't really remember exactly why but I think there was an issue with infinite render-loops when I did this, even with `useCallback`. I just remembered that when starting to change it like you said, adding the two dependencies here means I have to add lots of dependencies to those functions with `useCallback`.

I remember that no matter what I tried I got infinite re-render loops like this, I think the issue was that we have to add dependencies on object fields (e.g `data.chartDataDaily`) but I'm going to look into this again.

NicolaLS replied:
So I just tried it again and this time there are no problems at all, not sure what I did wrong last time.

NicolaLS commented on frontends/web/src/routes/account/summary/chart.tsx:
```diff
@@ -433,45 +425,36 @@ export const Chart = ({
+    if (!chartInitialized.current) {
+      initChart();
     }
+  });
```
Yes I intentionally added this so it stays the same as it currently is:
https://github.com/BitBoxSwiss/bitbox-wallet-app/blob/fb2888332c16bcc2b92f574b268db63b0d085c54/frontends/web/src/routes/account/summary/chart.tsx#L110-L114

I guess we can remove it, it should not be possible for a variable to become `undefined`/`null` if it is not explicitly assigned to it. So it should be enough to initialize it once with the other `useEffect`.

NicolaLS replied:
*also should never be set to false if we don't explicitly assign false

NicolaLS replied:
Adding the `initChart` and `removeChart` to the dependency array of the other `useEffect` and then removing this one works.

NicolaLS commented:
@shonsirsha thanks, adding `initChart`/`removeChart` and using `useCallback` fixed the chart so that the other `useEffect` that ran on every render could be removed. I did this in another commit but let me know if I should squash all three into one, only the last into the second one before merge.

thisconnect commented:
@NicolaLS please squash and rebase

thisconnect reviewed (changes requested):
lgtm with nit, also please reabase

NicolaLS commented:
@thisconnect first push is rebase without changes but fixing merge conflicts, I basically overwrote these commits:
https://github.com/BitBoxSwiss/bitbox-wallet-app/commit/c62a5c289d5b60aa9092e8c94bbccf157708dcbb#diff-c761d5505dba4edd4496dfe892f927231782977bae81a155d645a6a04dd7ab59L525
and
https://github.com/BitBoxSwiss/bitbox-wallet-app/commit/97c80ba63f9e1bb71f252da65b5fc838d552997b#diff-c761d5505dba4edd4496dfe892f927231782977bae81a155d645a6a04dd7ab59R141

But it looks like the new version of `chart.tsx` already fixes/addresses what these commits did.

Second push changed the nit.

thisconnect reviewed (approved):
tested LGTM thanks 👍
